### DISCUSSION TITLE
swfchan wishlist

### DISCUSSION THREAD

**Discussion by QualityLow** [ruffle-rs/ruffle#1743] [Ideas]
Originally planned to have this in my original thread at https://github.com/ruffle-rs/ruffle/discussions/1742 but thought that maybe it would be more suitable to have it in the Ideas category. Some of these are not really new ideas but others are.

My biggest wish list for Ruffle features are:

1. **Being able to get PercentLoaded from the JavaScript Ruffle player.**

Having access to the whole ExternalInterface (available in AS2) would be nice but being able to check the status of the swf download would be particularly useful. Right now there's no indication that the flash is being downloaded and together with not knowing if Ruffle can actually play the flash it leaves users staring at a blank space for a long time without knowing if the issue is with Ruffle or if it's the server being slow.

2. **Being able to get _currentframe and _framesloaded from the JavaScript Ruffle player.**

Those are usually received with ExternalInterface's GetVariable() but having access to them before ExternalInterface is implemented would allow a seekbar to be constructed for the flash, which is something people have gotten used to on swfchan. It will be a readonly seekbar until Ruffle makes GotoFrame() available from external JavaScript calls but a readonly seekbar is better than no seekbar. I find watching a 10 minute flash animation is a better experience if you can see how much is left.

3. **Glow filter.**

A lot of text in flash make use the glow filter. The blur and drop shadow filters would also be nice to have but the lack of glow is what I miss the most in terms of filters. Together with "knockout" and "inner glow" you could make things a lot more visually pleasing.

4. **Embedded video playback.**

There's a large amount of flashes that use embedded videos, especially on 4chan's /f/ board (which has been a focus of swfchan's archiving efforts throughout the years). I know licensing issues may have complicated implementing video codec support but it's still high on my wishlist.

5. **Fake basic ActionScript 3 support.**

Being able to handle simple things like play()/stop()/gotoAndPlay() in AS3 would go a long way, until AS3 support is implemented perhaps it's possible to scan the swf when it is loaded and replace recognized simple AS3 code with AS2 that does the same thing? This would enable playback of many AS3 flashes that don't really use AS3 exclusive things.

6. **Framerate acceleration (or frameskipping) to sync with streaming audio.**

I've noticed that if Ruffle is left playing in a different tab the animation will go out of sync with the audio and not correct itself when you return to the tab. Maybe the framerate of the flash can be boosted to catch up to the frame where the audio is? More boost the further behind the animation is and a slowdown if the animation is somehow ahead.

7. **Built-in sitelock bypass in right-click menu.**

It would be nice if the right-click menu had a "fake being offline" or "act as if local" option (selecting it sets a JavaScript cookie to ensure it is still checked the next time Ruffle loads). What it would do is make URL checks in ActionScript return a dummy file:///c:/flash/filename.swf URL instead of the actual URL hosting the flash (filename.swf can still be the real swf name). This allows playback of many sitelocked flashes that do allow local playback. Being able to bypass sitelocks that don't allow local playback and only a specific domain would be trickier to implement since you'd need to figure out which if statements that should always return true or false.

8. **Being able to disallow networking for the flash.**

When displaying swf files using embed you could set allowNetworking to all/internal/none to disallow it to communicate with the outside world. It prevented APIs such as for example navigateToURL(). I don't think allowNetworking was used too often but I had it set to none as default on swfchan to increase privacy and also because the site was never built around loading external swf files anyway. If Ruffle does implement support for allowNetworking set to "none" it would be nice if a small red dot briefly appeared next to the cursor every time a unique URL request was blocked, to signal that something was blocked. Then the user could right-click and select "open blocked URL in new window" in the right-click menu (displaying the full URL when the option is hovered).

9. **Smoothing on/off in right-click menu.**

Intentional or not it's overall good that you smooth images by default in Ruffle since a lot of flash creators forgot to enable smoothing for their images or didn't even know about it. However a small percentage of flashes is designed around looking pixly. For those flashes it would be a nice feature to be able to turn off smoothing in Ruffle's right-click menu. Perhaps smoothing could come with three modes: Always/When Requested/Never. "Always" would be Ruffle's default and "When Requested" would use the original flash plugin behaviour (off by default and on when images are specifically set to use smoothing).

10. **Rendering outside the stage.**

Sometimes, especially in flashes from 4chan's /f/, things are placed outside the visuals of the flash as an easter egg or to provide additional information. For example music source are often placed below the stage. When the swf files are opened directly in a browser tab they are scaled by the flash plugin to fit the browser window. So if you made the window tall you could see the song source "below the visuals". If Ruffle's goal is to emulate flash it should also emulate being able to see what's outside the stage if the container that holds the Ruffle player changes in size to be tall or thin. This important to a lot of people.

11. **Zooming in the right-click menu.**

It's not super important but something I miss being able to do. Zooming in is useful to check if something was made with vectors or if it's a raster image. In some rare cases there are also easter eggs that you can't really see if you don't zoom in on a part of a flash. Btw if you later implement support for ActionScript that modifies the right-click menu please never allow the default right-click options to be removed (like they can be in Adobe's flash plugin).

12. **Tab for hotspots.**

Pressing tab to put a big yellow border around clickable hotspots in the flash is a very useful timesaver, especially if you are browsing through several different flashes with non-obvious hotspots in different places. Enter to send a left-mouse-down event to the hotspot and Shift+Tab to go to the previous hotspot instead of the next one.

13. **Selectable and copyable text.**

You should be able to click selectable text and press Ctrl+A and Ctrl+C to copy it (or mark a section of the text and copy just that). Selecting and copying text doesn't seem to work in Ruffle right now.

--

Here's an example related to [9] and [10]: http://eye.swfchan.com/flash.asp?id=199231&n=Sub+Pen.swf
(The flash is safe for work but the ads on the site are not.)

[9] The visuals are designed to be pixly but Ruffle forces the images to be smooth when scaled (click "Fit" below the Ruffle player to enlarge the container holding it).
[10] Clicking "Thinner" six times below the Ruffle player should reveal music source beneath the stage in the flash but it doesn't render (invisible text is detected since the cursor changes when you hover it).

**Comment by n0samu** [comment]
1. I had never heard of this function and it's a bit difficult to find documentation on it. I don't think Ruffle is able to track the download progress of SWFs right now (its splash screen loading bar only tracks the download progress of Ruffle's own WASM file, I think).
2. GetVariable and SetVariable would be the most difficult of the Flash plugin's JavaScript API functions to implement, IMO.
3. The Ruffle developers are currently working on support for "cache as bitmap" support, which will lay the groundwork for beginning to support filters.
4. Ruffle supports playing embedded videos now, and support for external FLV videos is nearly finished. Neither will support the H.264 codec for now though.
5. Basic AS3 support could be considered complete now.
6. The solution we settled on was to pause the Ruffle player when the tab is not focused.
7. Sitelock bypassing is not something the Ruffle developers want to support.
8. Support for `allowNetworking` was added as part of #10625, but the "red dot" convenience feature sounds pretty useful and is not something we considered before. I'd encourage you to make a dedicated issue for that feature request.
9. I think you can set the movie to low quality to disable smoothing now? Not totally sure about this one. We do support the quality setting now, which at least changes the MSAA value.
10. Ruffle supports the same `scale` settings as Flash. You could set `scale` to "noscale" and expand the Ruffle embed's width/height to show the parts outside the stage area, for example.
11. Zooming is not supported yet but I don't think that would be too hard to add.
12. Tab navigation is a pretty important feature that Ruffle is missing, but I have no idea if/when anyone is going to work on it.
13. Support for copying, pasting and selecting text was added in #11059 and #10920. Ruffle still doesn't support selecting text by clicking and dragging the mouse, though.
